spot_wrapper/calibration/spot_in_hand_camera_calibration.py

```diff
@@ -0,0 +1,291 @@
+        duration_sec: float = 1.0,
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        def grab_state_as_transform() -> np.ndarray:
+            robot_state = self.robot_state_client.get_robot_state()
+            origin_t_planning_frame = get_a_tform_b(
+                robot_state.kinematic_state.transforms_snapshot,
+                BODY_FRAME_NAME,
+                HAND_FRAME_NAME,
+            )
+            pose_transform = np.eye(4)
+            pose_transform[:-1, -1] = np.array(
+                [
+                    origin_t_planning_frame.x,
+                    origin_t_planning_frame.y,
+                    origin_t_planning_frame.z,
+                ]
+            ).reshape((3,))
+            pose_transform[:-1, :-1] = origin_t_planning_frame.rot.to_matrix()
+            return pose_transform
+
+        if origin_t_planning_frame is None:
+            initial_pose = grab_state_as_transform()
+        else:
+            initial_pose = origin_t_planning_frame
+
+        try:
+            opencv_camera_t_viewpoint = np.eye(4)
+            opencv_camera_t_viewpoint[:3, :3] = self.hand_t_opencv_camera_rot
+        except AttributeError:
+            raise ValueError("Must call _set_localization_param prior to invoking this method.")
+        new_pose = convert_camera_t_viewpoint_to_origin_t_planning_frame(
+            origin_t_planning_frame=initial_pose,
+            planning_frame_t_opencv_camera=opencv_camera_t_viewpoint,
+            opencv_camera_t_viewpoint=transform_offset,
+        )
+
+        new_pose_quat = math_helpers.Quat.from_matrix(new_pose[:-1, :-1])
+
+        new_arm_command = RobotCommandBuilder.arm_pose_command(
+            new_pose[0, 3],  # x
+            new_pose[1, 3],  # y
+            new_pose[2, 3],  # z
+            new_pose_quat.w,
+            new_pose_quat.x,
+            new_pose_quat.y,
+            new_pose_quat.z,
+            BODY_FRAME_NAME,
+            duration_sec,
+        )
+        if not use_body:
+            command_id = self.command_client.robot_command(new_arm_command)
+        else:
+            follow_arm_command = RobotCommandBuilder.follow_arm_command()
+            command = RobotCommandBuilder.build_synchro_command(follow_arm_command, new_arm_command)
+            command_id = self.command_client.robot_command(command)
+        block_until_arm_arrives(
+            self.command_client,
+            command_id,
+            timeout_sec=duration_sec * 2,
+        )
+
+        return (initial_pose, new_pose)  # second value is new value
+
+    def shutdown(self) -> None:
+        stow_arm_command = RobotCommandBuilder.arm_stow_command()
+        ready_id = self.command_client.robot_command(stow_arm_command)
+        block_until_arm_arrives(self.command_client, ready_id)
+        blocking_sit(self.command_client)
+        self.keep_alive_leave.shutdown()
+        logger.info("IT IS NOW SAFE TO RETAKE CONTROL OF SPOT, DONE TAKING PICS :)")
+
+    def _set_localization_param(
```

Comment:
Could these params and estimates be pulled from the sdk somehow and not need someone to know the resolution ahead of time and maybe even get the factory estimate?

Reply:
The estimated intrinsic parameters can likely be pulled through the spot sdk, as can the current resolution (I'm pretty sure that the ROS 2 driver does this for the image info).

However, as I discovered, what mode the resolution is in can very from robot to robot depending on who was using it prior (resolution settings for the gripper are preserved across sessions, but not reboots). @erosen-bdai mentioned to me that there was interest in calibrating these robots in more than one resolution, as they are starting to test different resolutions with Spot. I thought that it would be better to be explicit here what mode you'd like to calibrate in prior to the calibration for the sake of transparency, as well as that a certain camera matrix is being used for that mode, so that it's easier for people to recreate for other robots.

I'd be down to read the camera matrix from the sdk with something like https://dev.bostondynamics.com/protos/bosdyn/api/proto_reference.html#bosdyn-api-ImageSource-PinholeModel-CameraIntrinsics

Reply:
Fair, yeah I remember its pretty controllable for the user, just wasnt sure it was well known enough that people will know what resolutions they are in. I just forsee people missing that, it might make sense to not default the resolution and have it just error then? Might avoid people from debugging trying to figure out why the transform is not working when they are on 1080. Idk if it's necessary though, so maybe just leave it for now. If it becomes a problem then we'll deal with it. TLDR: sounds good lol

Reply:
Yeah my hope was that people would see the log output 


```
logger.info(f"Setting gripper camera resolution to be {resolution}")
self.gripper_camera_client.set_camera_params(gripper_camera_params)
logger.info(f"Resolution set succesfully to {resolution}")
````

or, they would see the resolution in the calibration results under ```image_dim```, as I changed that to be included as well 

```
more_samples:
  intrinsic:
    1:
      camera_matrix: [225.00078392263674, 0.0, 110.86282677818623, 0.0, 224.94278925448418,
        91.79054340465953, 0.0, 0.0, 1.0]
      dist_coeffs: [0.1155545450372007, -3.42732171749554, -0.00019769867434315104,
        0.00587874888380888, 6.888146131015899]
      image_dim: [171, 224]
    0:
      camera_matrix: [551.4763619125183, 0.0, 321.7514190148494, 0.0, 551.3475634115229,
        238.75241303249507, 0.0, 0.0, 1.0]
      dist_coeffs: [0.15548534943475067, -0.4228757660120961, 0.000278244641828995,
        -0.0007271750729640895, 0.3607249187014709]
      image_dim: [480, 640]
  extrinsic:
    1:
      0:
        R: [-0.0007766179784000971, -0.9999885102094119, -0.00473036083998684, 0.9999271180999354,
          -0.0007195612951641436, -0.012051585783637294, 0.01204804352886777, -0.004739375560488635,
          0.9999161879709836]
        T: [0.021332555818574267, 0.01598620111299191, -0.02306447610443222]
  run_param:
    num_images: 729
    timestamp: '2024-08-19 03:43:06'
    stereo_pairs:
    - [1, 0]
    photo_utilization_ratio: 2
    num_checkers_width: 9
    num_checkers_height: 4
    dict_size: DICT_4X4_50
    checker_dim: 0.115
    marker_dim: 0.09
    data_path: /projects/spot_ros2/moana_25
    tag: more_samples
    unsafe_tag_save: false
```

Leaving as is for now although I'm down to take out the default / read the camera mat off of the APi